New Pull Request to merge AWSEMF exporter changes for EKS Fargate Container Insights

Aditya-Gollapudi opened the pull request:
Made a new pull request after rebasing the code, old pull request is here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3882

Added the option to have a user who is sending metrics from EKS Fargate Container Insights to reformat them to look the same as insights from ECS so that they can be ingested by CloudWatch.

Link to tracking Issue: No issue number to my knowledge

Testing: Still working on writing unit tests, but all unit tests will likely be with regards to the AddKubernetesWrapper function

Documentation: There is no added documentation right now, but the ADOT collector will tell customer to set the create_high_level_object flag to true in their awsemf exporter config (the flag defaults to false) and add "kubernetes" to the parse_json_encoded_attribtues array for the same exporter (if they don't do this the value for the "kubernetes" key will be the JSON encoding of the high level object rather than the object itself)

They will also have to use metrics transform to rename metrics to match the ECS labels (e.g. host_name rather than host.name) this will allow it to be used for other features in the future and to add Type = Container or Type = Pod as neccessary.

Aditya-Gollapudi committed:
Reintroduced changes builds and runs

Aditya-Gollapudi committed:
Added Dockerfile

Aditya-Gollapudi committed:
added some silly changes and added containers to the specialized section

Aditya-Gollapudi committed:
Removed commented out code and reset names to match ECS Insights

Aditya-Gollapudi committed:
Changed code back to building from lib

Aditya-Gollapudi committed:
Delete retest_config.yaml

Removed a local testing file

Aditya-Gollapudi committed:
Changed the name of the flag to be more descriptive and added some commenting

Aditya-Gollapudi committed:
Changed the error logging sop that it propogated up rather than simply fatally logging

Aditya-Gollapudi committed:
changed variable names of pod_template_hash and isPod

Aditya-Gollapudi committed:
readded container metric specfic values

Aditya-Gollapudi committed:
Removed some stray testing code

Aditya-Gollapudi committed:
I had a typo

Aditya-Gollapudi committed:
cleaned up the if statement

Aditya-Gollapudi committed:
Added a unit test

anuraaga commented on exporter/awsemfexporter/grouped_metric.go:
```diff
@@ -17,6 +17,8 @@ package awsemfexporter
 import (
 	"go.opentelemetry.io/collector/model/pdata"
 	"go.uber.org/zap"
+	"encoding/json"
```
Think these aren't grouped correctly and other files have wrong ordering, run the formatting on the source to fix the build

anuraaga commented on exporter/awsemfexporter/grouped_metric.go:
```diff
@@ -83,6 +97,130 @@ func addToGroupedMetric(pmd *pdata.Metric, groupedMetrics map[interface{}]*Group
+}
+
+type kubernetesObj struct{
+	Container_name string `json:`
```
Use standard go camel casing and the struct tag for the actual json name

anuraaga commented on exporter/awsemfexporter/grouped_metric.go:
```diff
@@ -83,6 +97,130 @@ func addToGroupedMetric(pmd *pdata.Metric, groupedMetrics map[interface{}]*Group
+
+
+func recursivelyFillInStruct(labels map[string]string, schema interface{}) (string, error){
+	jsonBytes, err := json.Marshal(schema)
```
I haven't looked into deeply how to improve it, but marshalling and unmarshalling and remarshalling seems like too much marshalling that you'd expect in code like this, can it be improved?

Aditya-Gollapudi replied:
This can be improved fairly easily by just submitted the schema as a map in the first place which would be a fairly simple change. Alternatively I could convert from a struct to a map using reflects but those are a lot less safe (there are also some external packages that wrap this functionality in a single command but introducing more dependencies also seemed like a poor idea)

The reason I set up the code to accept schemas as structs was I figured that would be more future proof but I am very new to the community to if that isn't the case I'm happy to switch the way schemas are submitted.

As for using a struct directly as a schema this is super messy for type related reasons and is a bad idea in general given the way Go handles reflection.

mxiamxia replied:
This code is messy. It looks your `schema/kubernetesObj` has all hardcoded value, why do you need do the recursively conversion?

Aditya-Gollapudi replied:
The reason to have the code be written in a generic way is so that it will be easy to change the names of incoming metrics or the structure of the kubernetes object in the future. I had considered at some point having the schema be given as input in the config but it was eventually decided against. Is there a specific aspect of the code you find messy or is it just the recursion over all.

Aditya-Gollapudi replied:
@mxiamxia @anuraaga @alolita @Aneurysm9 I believe I've dealt with all of the quick issues, if my answer to the above question was unsatisfying or you have any other concerns please let me know - otherwise hopefully we can get this PR merged soon

anuraaga replied:
Isn't the schema static? Why do we need a schema instead of just writing it as standard data translation code? This seems to be on the hot path so if we can accomplish the same with more direct code it's better.

Aditya-Gollapudi replied:
@anuraaga we made it non-static to make it easy to make changes in the future and to have it do more kinds of reformatting in the future. Because we are connecting two services we don't control we wanted to have some flexibility

Aneurysm9 replied:
> Isn't the schema static? Why do we need a schema instead of just writing it as standard data translation code? This seems to be on the hot path so if we can accomplish the same with more direct code it's better.

Agreed.  This feels like it should be a matter of filling in the appropriate labels.  If we find at a later date that a general purpose label translation/transformation framework is required here then we can build it at that point.

Aditya-Gollapudi replied:
@Aneurysm9  My main objections to the full hardcoding approach is  1) that in order to distinguish between a lack of a key and an empty string you would need a large number of if statements in a row creating unnecessary code complexity and making things harder to maintain 2) that it make it harder to make changes in the future if there are some slight changes in the format (as mentioned above).

Additionally, given that this approach has the same or lower maintenance costs than the hardcoded solution reverting to a purely hardcoded solution seems to be only a loss.

Aditya-Gollapudi committed:
Changed to camel case

Aditya-Gollapudi committed:
Changed to camel case

Aditya-Gollapudi committed:
Changed the code to accept a generic schema in the config

Left the existing code in their for now in case I want to switch back later

bogdandrutu commented:
@Aditya-Gollapudi please make sure your account is correctly setup and you have signed the CLA.

Aditya-Gollapudi committed:
Changed it back to being hard coded

Aditya-Gollapudi committed:
Removed a typo

mxiamxia commented on exporter/awsemfexporter/getJsonString:
```diff
@@ -0,0 +1,60 @@
+import{
```
This file is not a `golang` file.

Aditya-Gollapudi replied:
Oh wait that shouldn't be there, I'll remove it once I'm back at my work device

mxiamxia commented on exporter/awsemfexporter/getJsonString:
```diff
@@ -0,0 +1,60 @@
+
+type internalLabelsObj struct{
+	App string `json: "app"`
+	PodTemplateHash string `json: "pod-template-hash"`
```
:s/-/_/g

Aditya-Gollapudi replied:
I'm sorry, I think I'm unfamiliar with this notation. What does this mean?

Aneurysm9 replied:
Substitute '-' with '_', globally.

Aditya-Gollapudi replied:
Thanks!

mxiamxia commented on exporter/awsemfexporter/getJsonString:
```diff
@@ -0,0 +1,60 @@
+}
+
+
+func main(){
+	//create schema
+	schema := kubernetesObj{}
+	schema.ContainerName = "container_name"
+	schema.Docker =
+	internalDockerObj{
+		ContainerId: "container_id",
+	}
```
is this a test file?

Aditya-Gollapudi replied:
This was a mistake, I have since removed it

mxiamxia commented on exporter/awsemfexporter/config.go:
```diff
@@ -64,6 +64,12 @@ type Config struct {
 	OutputDestination string `mapstructure:"output_destination"`
 
+
+	// CreateEKSFargateKubernetesObject is an option to reformat certin metric labels so that they take the form of a high level object
+	// The end result will make the labels look like those coming out of ECS and be more easily injected into cloudwatch
```
I can't understand what does this new config do from the description. What does 'high level object' refer to?

Also, the config naming of  'CreateEKSFargateKubernetesObject' is also weird. It doesn't look like an attribute for Config.

mxiamxia replied:
where does `ParseJSONEncodedAttributeValues` being used in your changes?

Aditya-Gollapudi replied:
ParseJSONEncodedAttributeValues is being used implicitly as the code I added to the exporter creates a new label "kubernetes" and sets the value to the JSON string of the created object. We had considered having it so that setting the CreateEKSFargateKubernetesObject flag would automatically add "kubernetes" to ParseJSONEncodedAttributeValues but decided it would go against the design philosophy.

The high level object is just the way I was referring the the nested kubernetes object because it will be a high-level key-value pair. The flag simply tells the exporter whether or not to add in the new object. As for the naming of the flag it was just what made the most sense to me - I'm happy to rename it if something else is preferred.

anuraaga replied:
Do we need an option for this? If cloudwatch handles kubernetes objects better with this format we may as well just always format right? Alternatively, have you considered improving the backend instead of adding more complexity to clients such as the collector? Perhaps it's a good first step on the way to native OTLP support in Cloudwatch

Aditya-Gollapudi replied:
@anuraaga @hossain-rayhan We decided against doing this by default in case some customers had backwards compatibility issues with having the extra key in the labels map. I'm not certain why the feature was implemented here rather than in CW or EKS, Rayhan might know more.

hossain-rayhan replied:
Hi @anuraaga , I think we don't want to make this a default behavior because-
- This feature is very specific to EKS Fargate customers. For EKS EC2 customers, they will have the special metadata (`kubernetes`) with the metric itself and that would be done from the receiver. Also this exporter is being used by other customers like ECS and Batch (maybe) where we don't need this. 
- Having a flag will give the customers an option to opt-in which is good in my opinion. We may have some EKS Fargate customers who don't want Container Insights experience but want to ship metrics to CW. This flag will be helpful to support both use cases. 

For @mxiamxia's comment, can we rename the flag like, `eks_fargate_container_insights_enabled`? This name would be generic and we can apply this for other changes specific to EKS Fargate Container Insights (if needed).

mxiamxia commented on exporter/awsemfexporter/grouped_metric.go:
```diff
@@ -83,6 +97,138 @@ func addToGroupedMetric(pmd *pdata.Metric, groupedMetrics map[interface{}]*Group
+
+type internalLabelsObj struct{
+	App string `json:"app"`
+	PodTemplateHash string `json:"pod-template-hash"`
```
ditto

mxiamxia replied:
use "_" for jason attribute name since you had it everywhere else

Aditya-Gollapudi replied:
@mxiamxia unfortunately I can't do this as I need to match the format which cloud watch expects

mxiamxia commented on exporter/awsemfexporter/grouped_metric.go:
```diff
@@ -52,7 +54,19 @@ func addToGroupedMetric(pmd *pdata.Metric, groupedMetrics map[interface{}]*Group
+			if((metricType == "Pod" || metricType == "Container")&& config.CreateEKSFargateKubernetesObject){
+				 err := addKubernetesWrapper(labels)
+				if err != nil{
+					logger.Warn("Issue forming Kubernetes Object", zap.Error(err))
```
log description starts with lowercase

Aditya-Gollapudi committed:
Removed stray file

Didn't realize that renaming a file would make a copy

Aditya-Gollapudi committed:
Changed log warning to lower case

Aditya-Gollapudi committed:
Ran go fmt

Aditya-Gollapudi committed:
made Id into ID

Aditya-Gollapudi committed:
Ran goimports

Aditya-Gollapudi committed:
Another go fmt issue

Aditya-Gollapudi committed:
fixed some more formatting

Aneurysm9 commented on exporter/awsemfexporter/grouped_metric.go:
```diff
@@ -83,6 +97,135 @@ func addToGroupedMetric(pmd *pdata.Metric, groupedMetrics map[interface{}]*Group
+
+func addKubernetesWrapper(labels map[string]string) error {
+	//create schema
+	schema := kubernetesObj{}
+	schema.ContainerName = "container"
+	schema.Docker =
+		internalDockerObj{
+			ContainerID: "container_id",
+		}
+	schema.Host = "NodeName"
+	schema.Labels =
+		internalLabelsObj{
+			App:             "app",
+			PodTemplateHash: "pod-template-hash",
+		}
+	schema.NamespaceName = "Namespace"
+	schema.PodID = "PodId"
+	schema.PodName = "PodName"
+	schema.PodOwners =
+		internalPodOwnersObj{
+			OwnerKind: "owner_kind",
+			OwnerName: "owner_name",
+		}
+	schema.ServiceName = "Service"
```
```suggestion
	schema := kubernetesObj{
		ContainerName: "container",
		Docker: internalDockerObj{
			ContainerID: "container_id",
		},
		Host: "NodeName",
		Labels: internalLabelsObj{
			App:             "app",
			PodTemplateHash: "pod-template-hash",
		},
		NamespaceName: "Namespace",
		PodID: "PodId",
		PodName: "PodName",
		PodOwners: internalPodOwnersObj{
			OwnerKind: "owner_kind",
			OwnerName: "owner_name",
		},
		ServiceName: "Service",
	}
```

It is idiomatic to use the struct literal initialization form.

Aneurysm9 reviewed (changes requested):
I would not be comfortable landing this as it is.  The apparent unnecessary complexity needs to be removed or justified.

Aditya-Gollapudi committed:
Changed to struct initialization form

Co-authored-by: Anthony Mirabella <a9@aneurysm9.com>

Aditya-Gollapudi committed:
Fixed casing after rebase

Aditya-Gollapudi committed:
Changed to hardcoded approach and renamed flag

Aditya-Gollapudi commented:
@Aneurysm9 @alolita  @mxiamxia Sorry for the late change - I got preoccupied with some other work. I changed it to a hardcoded solution and changed the flag name. We tentatively decided to continue to make customers change both the flag and add the "kubernetes" key to ParseJSONAttributeValues because to do otherwise would involve us editing the config in code and we weren't sure whether that was allowed from a design philosophy perspective.

If you have any issues with the last decision or any other comments please let me know

Aditya-Gollapudi committed:
Changed something that lint complained about

anuraaga commented on exporter/awsemfexporter/grouped_metric.go:
```diff
@@ -83,6 +92,82 @@ func addToGroupedMetric(pmd *pdata.Metric, groupedMetrics map[interface{}]*group
+
+func addKubernetesWrapper(labels map[string]string) {
+	//fill in obj
+	filledInObj := kubernetesObj{
```
Way simpler :-)

Aneurysm9 commented on exporter/awsemfexporter/config.go:
```diff
@@ -64,6 +64,11 @@ type Config struct {
 	// TODO: we can support directing output to a file (in the future) while customer specifies a file path here.
 	OutputDestination string `mapstructure:"output_destination"`
 
+	// CreateEKSFargateKubernetesObject is an option to reformat certin metric labels so that they take the form of a high level object
```
```suggestion
	// EKSFargateContainerInsightsEnabled is an option to reformat certin metric labels so that they take the form of a high level object
```

Ensure the rest of this comment matches the current implementation.

Aneurysm9 commented on receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/rest_client.go:
```diff
@@ -31,6 +31,7 @@ func NewRestClient(client Client) *HTTPRestClient {
 }
 
 // EndpointResponse gets the task metadata and docker stats from ECS Task Metadata Endpoint
+
```
```suggestion
```

Aditya-Gollapudi committed:
Update exporter/awsemfexporter/config.go

Co-authored-by: Anthony Mirabella <a9@aneurysm9.com>

Aditya-Gollapudi committed:
Update receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/rest_client.go

Co-authored-by: Anthony Mirabella <a9@aneurysm9.com>

Aditya-Gollapudi commented:
@kbrockhoff would you mind looking at the code whenever you get the chance and running the unittest, windows-test, lint, and integration-tests (I can't run them myself as a first time contributor)

bogdandrutu commented:
@Aditya-Gollapudi please rebase and make sure that lint and unittests are passing
